internal/services/v1/lookupwatch.go

```diff
@@ -0,0 +1,303 @@
+package v1
+
+import (
+	"context"
+	"errors"
+
+	"github.com/authzed/spicedb/pkg/zedtoken"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	dispatchpkg "github.com/authzed/spicedb/internal/dispatch"
+	"github.com/authzed/spicedb/internal/graph"
+	datastoremw "github.com/authzed/spicedb/internal/middleware/datastore"
+	"github.com/authzed/spicedb/internal/middleware/handwrittenvalidation"
+	"github.com/authzed/spicedb/internal/middleware/usagemetrics"
+	"github.com/authzed/spicedb/internal/namespace"
+	"github.com/authzed/spicedb/internal/services/shared"
+	"github.com/authzed/spicedb/pkg/datastore"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+	dispatchv1 "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+	v1lookupwatch "github.com/authzed/spicedb/pkg/proto/lookupwatch/v1"
+	grpcmw "github.com/grpc-ecosystem/go-grpc-middleware"
+	grpcvalidate "github.com/grpc-ecosystem/go-grpc-middleware/validator"
+	"github.com/shopspring/decimal"
+)
+
+// NewLookupWatchServer creates a LookupWatchServiceServer instance.
+func NewLookupWatchServer(
+	dispatch dispatchpkg.Dispatcher,
+	defaultDepth uint32,
+) v1lookupwatch.LookupWatchServiceServer {
+	return &lookupWatchServer{
+		dispatch:     dispatch,
+		defaultDepth: defaultDepth,
+		WithServiceSpecificInterceptors: shared.WithServiceSpecificInterceptors{
+			Unary: grpcmw.ChainUnaryServer(
+				grpcvalidate.UnaryServerInterceptor(),
+				handwrittenvalidation.UnaryServerInterceptor,
+				usagemetrics.UnaryServerInterceptor(),
+			),
+			Stream: grpcmw.ChainStreamServer(
+				grpcvalidate.StreamServerInterceptor(),
+				handwrittenvalidation.StreamServerInterceptor,
+				usagemetrics.StreamServerInterceptor(),
+			),
+		},
+	}
+}
+
+type lookupWatchServer struct {
+	v1lookupwatch.UnimplementedLookupWatchServiceServer
+	shared.WithServiceSpecificInterceptors
+
+	dispatch     dispatchpkg.Dispatcher
+	defaultDepth uint32
+}
+
+func (lw *lookupWatchServer) WatchAccessibleResources(req *v1lookupwatch.WatchAccessibleResourcesRequest, stream v1lookupwatch.LookupWatchService_WatchAccessibleResourcesServer) error {
+	ctx := stream.Context()
+
+	ds := datastoremw.MustFromContext(ctx)
+	objectTypesMap := make(map[string]struct{})
+	for _, objectType := range []string{req.ResourceObjectType} {
+		objectTypesMap[objectType] = struct{}{}
+	}
+
+	var afterRevision decimal.Decimal
+	if req.OptionalStartTimestamp != nil && req.OptionalStartTimestamp.Token != "" {
+		decodedRevision, err := zedtoken.DecodeRevision(req.OptionalStartTimestamp)
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "failed to decode start revision: %s", err)
+		}
+
+		afterRevision = decodedRevision
+	} else {
+		var err error
+		afterRevision, err = ds.OptimizedRevision(ctx)
```

Comment:
Probably want to start this at `HeadRevision`, so you start at the most recent place and watch from there

Reply:
Done, thanks !